pandas/core/indexes/range.py

```diff
@@ -1037,3 +1038,43 @@ def _arith_method(self, other, op):
         except (ValueError, TypeError, ZeroDivisionError):
             # test_arithmetic_explicit_conversions
             return super()._arith_method(other, op)
+
+    def take(
+        self,
+        indices,
+        axis: Axis = 0,
+        allow_fill: bool = True,
+        fill_value=None,
+        **kwargs,
+    ):
+        if kwargs:
+            nv.validate_take((), kwargs)
+        if is_scalar(indices):
+            raise TypeError("Expected indices to be array-like")
+        indices = ensure_platform_int(indices)
+        allow_fill = self._maybe_disallow_fill(allow_fill, fill_value, indices)
+        assert allow_fill is False, "allow_fill isn't supported by RangeIndex"
```

Comment:
i dont think this is ensured by _maybe_disallow_fill.  we shouldn't be raising AssertionError for users

Reply:
I don't see a way to get `True` from [_maybe_disallow_fill](https://github.com/pandas-dev/pandas/blob/8cf4ab491b06274f555d1ea305c172354208747a/pandas/core/indexes/base.py#L1169) when using RangeIndex, because it calls [_can_hold_na](https://github.com/pandas-dev/pandas/blob/8cf4ab491b06274f555d1ea305c172354208747a/pandas/core/indexes/base.py#L2196) that always returns `False`, since [RangeIndex.dtype](https://github.com/pandas-dev/pandas/blob/8cf4ab491b06274f555d1ea305c172354208747a/pandas/core/indexes/range.py#L311) is fixed to int64.

Maybe would it be clearer if we just explicitly raise the exception in the subclass method, although there would be some redundancy with the exception raised in _maybe_disallow_fill?

i.e. removing the call to `_maybe_disallow_fill` and the assert, and add here:

```python
        if allow_fill and fill_value is not None:
            cls_name = type(self).__name__
            raise ValueError(
                f"Unable to fill values because {cls_name} cannot contain NA"
            )
```